### ISSUE TITLE
Transitive dependencies have wrong restriction

### ISSUE THREAD

**Issue by csmager** [fsprojects/Paket#3232]
This seems to be similar to #1213. 

I have a web app that I've split into two projects so the EF stuff can be in C# and the rest in F#. 

| Project | Target | Package Ref |
| --- | --- | --- |
| Web App | `netcoreapp2.1` | `Microsoft.AspNetCore.App` |
| EF Lib | `netstandard2.0` | `Microsoft.EntityFrameworkCore.SqlServer` |

So the initial `paket.dependencies` looks like this:
```
source https://api.nuget.org/v3/index.json

nuget Microsoft.AspNetCore.App
nuget Microsoft.EntityFrameworkCore.SqlServer
```

The `paket.lock` file ends up with this for the EF package:
```
Microsoft.EntityFrameworkCore.SqlServer (2.1)
  Microsoft.EntityFrameworkCore.Relational (>= 2.1) - restriction: >= netstandard2.0
  System.Data.SqlClient (>= 4.5) - restriction: >= netstandard2.0
```

As the EF bits are also pulled in by the meta-package, I would have thought (and `paket simplify` agrees) that `Microsoft.EntityFrameworkCore.SqlServer` can be removed from `paket.dependencies`. If I do this, remove `paket.lock` and run `paket install` again, then a bunch of restrictions change. The previous excerpt becomes this:
```
Microsoft.EntityFrameworkCore.SqlServer (2.1) - restriction: >= netcoreapp2.1
  Microsoft.EntityFrameworkCore.Relational (>= 2.1) - restriction: >= netstandard2.0
  System.Data.SqlClient (>= 4.5) - restriction: >= netstandard2.0
```
And now the EF project won't build because there's no package available for `netstandard2.0`.

I've created a [minimal repro solution](https://github.com/csmager/paket-transitives-repro) where you can see the [diff between the `paket.lock` files](https://github.com/csmager/paket-transitives-repro/compare/876696de01488130d19348a988ba1ed92a4cfe91...master#diff-2bfb11ab4230093400363f8d70eab088).

**Comment by matthid** [comment]
`paket simplify` is only a heuristic and gets things wrong in these scenarios (and simplify considering restrictions is quite a complicated topic).

The other issue is very similar or in fact identical to the reasoning here: https://github.com/fsprojects/Paket/issues/3082#issuecomment-368957508

**Comment by csmager** [comment]
Yep, it's the same as that. So we think this behaviour is correct, then?

If that's the case, what might be useful is that you get some sort of warning on restore when you have a package in `paket.references` for a project but that package reference couldn't be satisfied.

As it is, the dependency does exist in `paket.lock` (so you don't get an error), but it's restricted so it can't be used even though the package doesn't have those restrictions. It's a bit confusing!

**Comment by matthid** [comment]
Yes the behavior is correct. And yes a warning might be reasonable if people add references to projects were the target-frameworks indicate that it is filtered out. I would assume that this might potentially lead to false positives for the warning as it is a bit tricky to get correct

**Comment by csmager** [comment]
Yes, I'd also not considered the situation where a project has multiple target frameworks. It may get complicated.

**Comment by BlythMeister** [comment]
I wonder if https://github.com/fsprojects/Paket/pull/3523 might fix this...I've made it so a simplify will not remove a dependency if it has version rules or settings.

**Comment by BlythMeister** [comment]
Short of just adding the dependency to the paket.dependencies file such and don't let simply remove it (for example adding `>=0` as a version range once #3523 is merged.
Is there no way we can make the simplify not remove something that will actually change the output in the lock file?

Maybe each thing that is removed as part of a simplify in a dependencies file rebuilds the paket.lock to ensure it didn't change anything?

**Comment by BlythMeister** [comment]
As part of #3523 I am going to add logic that if the dependencies line has 'no-simplify' in a comment, the simplify command will not remove the item.
This will mean you can prevent the simplify from breaking your lock file.  It's not a fix, but a workaround
